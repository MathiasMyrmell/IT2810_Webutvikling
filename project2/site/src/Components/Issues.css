.Issues {
    width: 1150px;
    margin: 0 auto;
    padding: 25px;
    word-wrap: break-word;
}

.Issues>h1 {
    font-size: clamp(24px, 4vw, 50px);
}

.Issues>p {
    font-size: clamp(14px, 1.5vw, 20px);
}

/* For small screens */
@media (max-width: 1200px) {
    .Issues {
        width: 90%;
    }
}

/* For mobile phones */
@media (max-width: 512px) {
    .Issues {
        padding: 0;
    }
}