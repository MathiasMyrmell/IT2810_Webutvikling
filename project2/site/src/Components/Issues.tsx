import React from 'react';
import './Issues.css';
import { useEffect, useState, useContext, createContext } from "react";
import IssuesList from './IssuesList';
import './IssuesList.css';

export type IssuesContextState = {
    open: Array<Object>;
    closed: Array<Object>;
};

// For initializing
const issuesDefaultValues: IssuesContextState = {
    open: [],
    closed: [],
};
export const IssuesContext = createContext<IssuesContextState>(issuesDefaultValues);

function Issues() {
    const [repoData, setRepoData] = useState<IssuesContextState>(issuesDefaultValues);

    let url: string = localStorage.getItem("url") || "";
    let token: string = sessionStorage.getItem("token") || "";

    // On mount
    useEffect(() => {
        repoDataURL();
    }, []);

    async function repoDataURL() {
        // Get repo data about github user
        const api: string = ".no/api/v4/projects/";
        let urlArray: Array<string> = url.split(".no/");

        let openIssues: Array<Object> = [];
        let closedIssues: Array<Object> = [];
        await fetch(urlArray[0] + api + urlArray[1].replaceAll("/", "%2F") + "/issues" + "?private_token=" + token)
            .then((res) => res.json())
            .then(
                (result) => {
                    result.map((item: any) => {
                        if (item.closed_by == null) { // If the issue is open or not
                            openIssues.push({ id: item.iid, title: item.title });

                        } else {
                            closedIssues.push({ id: item.iid, title: item.title });

                        }
                    });
                    var issuesValues: IssuesContextState = {
                        open: openIssues,
                        closed: closedIssues,
                    };
                    setRepoData(issuesValues);
                },
                (error) => {
                    console.log(error);
                }
            );
    }
    if (repoData == issuesDefaultValues) { // Check if we have fetched yet
        return (
            <div className='Issues'>
                <h1>Visualisering/presentasjon av GITLAB data</h1>
                <p>Loading...</p>
            </div>
        );
    } else {
        return (
            <div className='Issues'>
                <h1>Visualisering/presentasjon av GITLAB data</h1>
                <p>Nå ser du på prosjektet med URL: {url}. På denne siden har vi brukt Context API for å videreføre GITLAB data.</p>
                <IssuesContext.Provider value={repoData}>
                    <IssuesList ></IssuesList>
                </IssuesContext.Provider>
            </div>
        );
    }
}

export default Issues;
