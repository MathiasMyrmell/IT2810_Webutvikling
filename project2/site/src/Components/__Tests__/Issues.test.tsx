import renderer from 'react-test-renderer';
import Issues from '../Issues';

test('Snapshot test Issues', () => {
    const tree = renderer.create(<Issues />).toJSON();
    expect(tree).toMatchSnapshot();
});
