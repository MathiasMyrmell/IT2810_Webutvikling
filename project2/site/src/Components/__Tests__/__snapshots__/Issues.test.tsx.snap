// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test Issues 1`] = `
<div
  className="Issues"
>
  <h1>
    Visualisering/presentasjon av GITLAB data
  </h1>
  <p>
    Loading...
  </p>
</div>
`;
